tests/lib/util/source-code.js

```diff
@@ -928,19 +988,488 @@ describe("SourceCode", () => {
+            eslint.on("CallExpression", assertCommentCount(0, 0));
+            eslint.on("BinaryExpression", assertCommentCount(0, 0));
+            eslint.on("Literal", assertCommentCount(0, 0));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return leading comments if the code only contains comments", () => {
+            const code = [
+                "//comment",
+                "/*another comment*/"
+            ].join("\n");
+
+            eslint.on("Program", assertCommentCount(2, 0));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return trailing comments if a block statement only contains comments", () => {
+            const code = [
+                "{",
+                "    //comment",
+                "    /*another comment*/",
+                "}"
+            ].join("\n");
+
+            eslint.on("Program", assertCommentCount(0, 0));
+            eslint.on("BlockStatement", assertCommentCount(0, 2));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return trailing comments if a class body only contains comments", () => {
+            const code = [
+                "class Foo {",
+                "    //comment",
+                "    /*another comment*/",
+                "}"
+            ].join("\n");
+
+            eslint.on("Program", assertCommentCount(0, 0));
+            eslint.on("ClassDeclaration", assertCommentCount(0, 0));
+            eslint.on("ClassBody", assertCommentCount(0, 2));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return trailing comments if an object only contains comments", () => {
+            const code = [
+                "({",
+                "    //comment",
+                "    /*another comment*/",
+                "})"
+            ].join("\n");
+
+            eslint.on("Program", assertCommentCount(0, 0));
+            eslint.on("ExpressionStatement", assertCommentCount(0, 0));
+            eslint.on("ObjectExpression", assertCommentCount(0, 2));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return trailing comments if an array only contains comments", () => {
+            const code = [
+                "[",
+                "    //comment",
+                "    /*another comment*/",
+                "]"
+            ].join("\n");
+
+            eslint.on("Program", assertCommentCount(0, 0));
+            eslint.on("ExpressionStatement", assertCommentCount(0, 0));
+            eslint.on("ArrayExpression", assertCommentCount(0, 2));
+
+            eslint.verify(code, config, "", true);
+        });
+
+        it("should return trailing comments if a switch statement only contains comments", () => {
+            const code = [
+                "switch (foo) {",
+                "    //comment",
+                "    /*another comment*/",
+                "}"
```

Comment:
Also wanted to mention that I'm proposing to actually deprecate `sourceCode.getComments()`: https://github.com/eslint/eslint/issues/8408

Ideally, my hope is that we can get `getComments()` in a place where it's good enough to not break existing rules currently relying on it. I don't think it's really feasible to make it have the exact same behavior as Espree's comment attachment strategy, and switching to thinking of comments in the context of token lists rather than AST nodes will make dealing with comments much more predictable!